### ISSUE TITLE
Support ECR authentication

### ISSUE THREAD

**Issue by hobbsh** [argoproj-labs/argocd-image-updater#112]
**Is your feature request related to a problem? Please describe.**
It looks like there is a lack of support for getting the API token required to authenticate against the ECR API. It would be great to be able to make use of the IAM Instance Profile assigned to a node which has `GetAuthorizationToken` permissions. It might look similar to this: https://github.com/awslabs/amazon-ecr-credential-helper/blob/master/ecr-login/api/client.go.

**Describe the solution you'd like**
The image updater supports ECR authentication

**Describe alternatives you've considered**
There do not seem to be any alternatives without code changes.

**Comment by hobbsh** [comment]
For additional context, this is the logged error I get when trying to use ECR:

```
time="2020-10-23T01:53:29Z" level=error msg="Could not get tags from registry: Get \"https://ACCOUNT_ID.dkr.ecr.us-west-2.amazonaws.com/v2/dev/my-image/tags/list\": http: non-successful response (status=401 body=\"Not Authorized\\n\")" alias=backend application=my-application image_name=dev/my-image image_tag=304dbc93e21898e95272d7cb81b61a671a8b7365 registry=ACCOUNT_ID.dkr.ecr.us-west-2.amazonaws.com
```

**Comment by hobbsh** [comment]
We can use `aws ecr get-authorization-token --output text --query 'authorizationData[].authorizationToken'` and jam it into a k8s secret (it yields a base64 encoded `username:password` string) however the token is only valid for 12 hours so there would need to be a cronjob that updates this secret every 12 hours.

**Comment by jannfis** [comment]
We currently have three methods for getting credentials right now - `secret`, `pullsecret` and `env`. Would a fourth option, that calls a script and re-uses its output as credentials, be sufficient for this use case?

For example we could introduce a new method `ext:/path/to/script.sh`, and you could either build your own image from the image updater's docker file or use an init container to copy the script to some location into the image updater's container.

**Comment by jannfis** [comment]
A feature implementing my above proposal has been merged with #121 and will be part of v0.8 release.

**Comment by jannfis** [comment]
I have no means to test against a real ECR instance, but would a script require some kind of parametrization in order to be able to retrieve credentials for the correct registry?

**Comment by vistrcm** [comment]
Hello, 
I can try to help with testing. ECR support is crucial for me too.

One question before I can test: do credentials cached somehow? I mean, does argocd-image-updater read credentials from the secret, env variable or execute the script every time or only once and then use these results?

It looks like function `SetEndpointCredentials` sets credentials for `RegistryEndpoint` and never update that
https://github.com/argoproj-labs/argocd-image-updater/blob/6723a25cfcd3f117f0d234b5449db7a6cc68bc2d/pkg/registry/registry.go#L168

Sorry, I had no chance to look deeper. 

If credentials are cached script will not help with ECR. By default, ECR tokens rotated every 12 hours.

**Comment by vistrcm** [comment]
It looks like credentials are cached if defined as a [parameter of registry configuration](https://argocd-image-updater.readthedocs.io/en/stable/configuration/registries/#specifying-credentials-for-accessing-container-registries) and not cached if specified [as annotation for the application.](https://argocd-image-updater.readthedocs.io/en/stable/configuration/images/#specifying-pull-secrets)

So using that knowledge, I was able to annotate my applications with
```yaml
annotations:
  argocd-image-updater.argoproj.io/image-list: org/app=XXXXXXXXXXXX.dkr.ecr.region.amazonaws.com/org/app
  argocd-image-updater.argoproj.io/org_app.update-strategy: latest
  argocd-image-updater.argoproj.io/org_app.kustomize.image-name: org/app
  argocd-image-updater.argoproj.io/org_app.pull-secret: secret:argocd-image-updater/aws-ecr-creds#creds
```

And create simple k8s CronJob to get token and update secret value:
```yaml
apiVersion: v1
kind: ServiceAccount
metadata:
  name: ecr-secret-udpater
---
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  name: ecr-secret-udpater
rules:
  - apiGroups:
      - ""
    resources:
      - secrets
    verbs:
      - update
---
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: ecr-secret-udpater
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: Role
  name: ecr-secret-udpater
subjects:
  - kind: ServiceAccount
    name: ecr-secret-udpater
---
apiVersion: v1
kind: Secret
metadata:
  annotations:
    description: this secret is dynamically updated by the k8s CronJob ecr-secret-update. store ECR registry user/token
  name: aws-ecr-creds
stringData:
  creds: will_be_set_by_the_job
---
apiVersion: batch/v1beta1
kind: CronJob
metadata:
  name: ecr-secret-update
spec:
  jobTemplate:
    spec:
      template:
        spec:
          containers:
            - args:
                - -c
                - kubectl create secret generic aws-ecr-creds --from-literal=creds=AWS:$(cat /store/token) --dry-run=client -o yaml | kubectl replace -f -
              command:
                - sh
              image: org/kubectl:v1.19.4
              name: kubectl
              volumeMounts:
                - mountPath: /store
                  name: store
          initContainers:
            - args:
                - -c
                - aws ecr get-login-password --region us-west-2 > /store/token
              command:
                - sh
              image: amazon/aws-cli:2.1.6
              name: get-login-password
              volumeMounts:
                - mountPath: /store
                  name: store
          restartPolicy: OnFailure
          serviceAccountName: ecr-secret-udpater
          volumes:
            - emptyDir:
                medium: Memory
              name: store
      ttlSecondsAfterFinished: 100
  schedule: '* */6 * * *'
```

This approach works for me on v0.7.0. I can share AWS related policies I used to grant permissions and details on org/kubectl:v1.19.4 image is someone is interested.

**Comment by jannfis** [comment]
@vistrcm Thank you for your input! So I see the problem with the cached credentials. 

I think an expiry time for the credentials might help in case of the credentials being cached on registry level, something like a new toggle `credentials_expire: <duration>` for the registry configuration. If credentials are older than `<duration>`, they'll be regenerated from the script.

I'm planning to release v0.8 this weekend hopefully, so such a change (if helpful) could make it in there, I believe.

**Comment by hobbsh** [comment]
I use the same solution currently to update ECR creds and it would be great if the image-updater could handle this itself.

**Comment by vistrcm** [comment]
> @vistrcm Thank you for your input! So I see the problem with the cached credentials.
> 
> I think an expiry time for the credentials might help in case of the credentials being cached on registry level, something like a new toggle `credentials_expire: <duration>` for the registry configuration. If credentials are older than `<duration>`, they'll be regenerated from its source (i.e. secret read again, or script executed again)
> 
> I'm planning to release v0.8 this weekend hopefully, so such a change (if helpful) could make it in there, I believe.

that would be great! thank you

**Comment by jannfis** [comment]
So with #124 merged, you can now specify an expiration time for your credentials, when configured at the registry level (i.e. in `registries.conf`). So now you can use something like the following:

```yaml
credentials: /some/where/eks-creds.sh
credexpire: 11h59m
```

The script at `/some/where/eks-creds.sh` should be a wrapper to call the `aws` CLI with all required parametrization, and output the resulting credentials as single line on stdout in the format `<username>:<token>`. As long as credentials are not expired (or image-updater is restarted), script will not be called and credentials are cached in-memory. After expiration time, the script will be called again to generate a new token.

Of course you would have to create an init container that copies required tools into the image updater's container.

Is this something you'll be able to use? I'll not be able to provide native ECR/AWS support, because I don't use AWS.

**Comment by hobbsh** [comment]
I'll try to test that in the next week or two. There's a typo in the docs, you mention `EKS` instead of `ECR` 😉

**Comment by jannfis** [comment]
I have released v0.8.0 today. To simplify testing, it introduced a new command `test` to the `argocd-image-updater` CLI. This command can also be used to test authentication to registries, without having to run it inside Kubernetes (and therefore possibly build a new image) and without the need to trial&error annotate your Argo CD applications.

You can simply test an authentication script as follows:

```shell
argocd-image-updater test <your_image_on_ecr> --credentials ext:/path/to/your/script --registries-conf /path/to/your/registries.conf
```

It can also do more to help you check image updater's behaviour for your specific images, and is rudimentary [documented here](https://argocd-image-updater.readthedocs.io/en/latest/install/start/#testing-beforehand)

The binary (only linux-amd64 so far) can be downloaded from the [release page](https://github.com/argoproj-labs/argocd-image-updater/releases/tag/v0.8.0)

**Comment by diranged** [comment]
Just for anyone who finds this and is wondering what the TLDR is ... this seems to work:

*registries.conf*:
```
registries:
- name: ECR
  api_url: https://xxx.dkr.ecr.us-west-2.amazonaws.com
  prefix: xxx.dkr.ecr.us-west-2.amazonaws.com
  ping: yes
  insecure: no
  tagsortmode: latest-first
  credentials: ext:/path/to/ecr.sh
  credsexpire: 10h
```

*ecr.sh*
```
#!/bin/sh
aws ecr --region us-west-2 get-authorization-token --output text --query 'authorizationData[].authorizationToken' | base64 -d
```

```
$ ./dist/argocd-image-updater  test xxx.dkr.ecr.us-west-2.amazonaws.com/dev/xyz-batch-serving --registries-conf ./registries.conf
INFO[0000] getting image                                 image_name=dev/xxx-batch-serving registry=xxx.dkr.ecr.us-west-2.amazonaws.com
DEBU[0000] rate limit for https://xxx.dkr.ecr.us-west-2.amazonaws.com is 2147483647 
INFO[0000] Loaded 1 registry configurations from ./registries.conf 
INFO[0000] git/argocd-image-updater/ecr.sh  dir= execID=c6FS4
INFO[0000] Fetching available tags and metadata from registry  image_name=dev/xyz-batch-serving
INFO[0001] Found 288 tags in registry                    image_name=dev/xxx-batch-serving
DEBU[0001] could not parse input tag abc-xyz as semver: Invalid Semantic Version 
...
```

**Comment by jannfis** [comment]
I will close this issue now, since I think the feature works now. Feel free to reopen it when you think there should be more work done to support ECR auth.

**Comment by jeroenmaas** [comment]
@diranged 

Thanks for sharing your config. It worked well, though one small note. I had to change _tagsortmode: latest-first_ to none since our ECR repo doesn't return the tags in order causing latest to not work when the value is specified.

**Comment by diranged** [comment]
> @diranged
> 
> Thanks for sharing your config. It worked well, though one small note. I had to change _tagsortmode: latest-first_ to none since our ECR repo doesn't return the tags in order causing latest to not work when the value is specified.

@jeroenmaas Can you share logs or more of your config showing that `tagsortmode: none` works for you? I just opened up https://github.com/argoproj-labs/argocd-image-updater/issues/216 because it turns out that I am seeing that fail on our side making the API calls to AWS.

**Comment by diranged** [comment]
Just following up - the actual issue was a permissions issue. See https://github.com/argoproj-labs/argocd-image-updater/issues/216#issuecomment-862382796.

**Comment by AndresJulia** [comment]
> Just for anyone who finds this and is wondering what the TLDR is ... this seems to work:
> 
> _registries.conf_:
> 
> ```
> registries:
> - name: ECR
>   api_url: https://xxx.dkr.ecr.us-west-2.amazonaws.com
>   prefix: xxx.dkr.ecr.us-west-2.amazonaws.com
>   ping: yes
>   insecure: no
>   tagsortmode: latest-first
>   credentials: ext:/path/to/ecr.sh
>   credsexpire: 10h
> ```
> 
> _ecr.sh_
> 
> ```
> #!/bin/sh
> aws ecr --region us-west-2 get-authorization-token --output text --query 'authorizationData[].authorizationToken' | base64 -d
> ```
> 
> ```
> $ ./dist/argocd-image-updater  test xxx.dkr.ecr.us-west-2.amazonaws.com/dev/xyz-batch-serving --registries-conf ./registries.conf
> INFO[0000] getting image                                 image_name=dev/xxx-batch-serving registry=xxx.dkr.ecr.us-west-2.amazonaws.com
> DEBU[0000] rate limit for https://xxx.dkr.ecr.us-west-2.amazonaws.com is 2147483647 
> INFO[0000] Loaded 1 registry configurations from ./registries.conf 
> INFO[0000] git/argocd-image-updater/ecr.sh  dir= execID=c6FS4
> INFO[0000] Fetching available tags and metadata from registry  image_name=dev/xyz-batch-serving
> INFO[0001] Found 288 tags in registry                    image_name=dev/xxx-batch-serving
> DEBU[0001] could not parse input tag abc-xyz as semver: Invalid Semantic Version 
> ...
> ```

But how do you get awscli inside the container?

**Comment by hobbsh** [comment]
> @diranged How do you install the script ecr.sh? How did you make it so that the argo-cd-image-updater pod can access the script? Thanks!

You will have to mount it as a configmap

**Comment by hobbsh** [comment]
> But how do you get awscli inside the container?

@AndresJulia I have not tried this but I am assuming people using this are extending the `argocd-image-updater` image to install `awscli`. Not ideal but that would work.

**Comment by sgavrylenko** [comment]
> 
> But how do you get awscli inside the container?

At this moment, awscli is present in the image

**Comment by joebowbeer** [comment]
IRSA is the most robust way to provide this permission, right? And there's nothing currently preventing using IRSA?

**Comment by diranged** [comment]
Sorry for the radio silence.. so here's an example of our image updater statefulset. To answer the questions, we put the `awscli` command in there by using an initcontainer. We get the ecr script in there through a mounted configmap.

```yaml
apiVersion: apps/v1
kind: StatefulSet
metadata:
  labels:
    app.kubernetes.io/component: image-updater
    app.kubernetes.io/instance: argocd-service
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/part-of: argocd
  name: argocd-service-image-updater
  namespace: argocd
spec:
  podManagementPolicy: OrderedReady
  replicas: 1
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      app.kubernetes.io/name: argocd-service-image-updater
  serviceName: argocd-service-image-updater
  template:
    metadata:
      annotations:
        config.value.checksum: 6b0afb78dbf20fea4bacf969b90124be13f4f583502bf014e5fa2371ab8d1acd
      creationTimestamp: null
      labels:
        app: argocd-service-image-updater
        app.kubernetes.io/name: argocd-service-image-updater
    spec:
      containers:
      - command:
        - argocd-image-updater
        - run
        env:
        - name: PATH
          value: /shared/bin:/opt/bitnami/aws-cli/venv/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
        - name: HOME
          value: /tmp
        - name: APPLICATIONS_API
          valueFrom:
            configMapKeyRef:
              key: applications_api
              name: argocd-service-image-updater-config
              optional: true
        - name: ARGOCD_GRPC_WEB
          valueFrom:
            configMapKeyRef:
              key: argocd.grpc_web
              name: argocd-service-image-updater-config
              optional: true
        - name: ARGOCD_SERVER
          valueFrom:
            configMapKeyRef:
              key: argocd.server_addr
              name: argocd-service-image-updater-config
              optional: true
        - name: ARGOCD_INSECURE
          valueFrom:
            configMapKeyRef:
              key: argocd.insecure
              name: argocd-service-image-updater-config
              optional: true
        - name: ARGOCD_PLAINTEXT
          valueFrom:
            configMapKeyRef:
              key: argocd.plaintext
              name: argocd-service-image-updater-config
              optional: true
        - name: ARGOCD_TOKEN
          valueFrom:
            secretKeyRef:
              key: argocd.token
              name: argocd-image-updater-secret
              optional: true
        - name: IMAGE_UPDATER_LOGLEVEL
          valueFrom:
            configMapKeyRef:
              key: log.level
              name: argocd-service-image-updater-config
              optional: true
        - name: GIT_COMMIT_USER
          valueFrom:
            configMapKeyRef:
              key: git.user
              name: argocd-service-image-updater-config
              optional: true
        - name: GIT_COMMIT_EMAIL
          valueFrom:
            configMapKeyRef:
              key: git.email
              name: argocd-service-image-updater-config
              optional: true
        image: public.ecr.aws/bitnami/aws-cli:latest
        imagePullPolicy: Always
        livenessProbe:
          failureThreshold: 3
          httpGet:
            path: /healthz
            port: 8080
            scheme: HTTP
          initialDelaySeconds: 3
          periodSeconds: 30
          successThreshold: 1
          timeoutSeconds: 1
        name: argocd-image-updater
        ports:
        - containerPort: 8080
          protocol: TCP
        readinessProbe:
          failureThreshold: 3
          httpGet:
            path: /healthz
            port: 8080
            scheme: HTTP
          initialDelaySeconds: 3
          periodSeconds: 30
          successThreshold: 1
          timeoutSeconds: 1
        resources: {}
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /app/config
          name: registries-conf
        - mountPath: /app/scripts
          name: ecr-login
        - mountPath: /shared
          name: shared
      dnsPolicy: ClusterFirst
      initContainers:
      - args:
        - -c
        - mkdir -p /shared/bin && cp /usr/local/bin/* /shared/bin
        command:
        - sh
        image: ...-2.amazonaws.com/vendor/argoprojlabs/argocd-image-updater:release-06-15-2021
        imagePullPolicy: Always
        name: argocd-image-updater-installer
        resources: {}
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /shared
          name: shared
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext:
        fsGroup: 1000
      serviceAccount: argocd-service-image-updater
      serviceAccountName: argocd-service-image-updater
      terminationGracePeriodSeconds: 30
      volumes:
      - configMap:
          defaultMode: 420
          items:
          - key: registries.conf
            path: registries.conf
          name: argocd-service-image-updater-config
          optional: true
        name: registries-conf
      - configMap:
          defaultMode: 493
          items:
          - key: ecr-login.sh
            path: ecr-login.sh
          name: argocd-service-image-updater-config
          optional: true
        name: ecr-login
      - emptyDir: {}
        name: shared
  updateStrategy:
    rollingUpdate:
      partition: 0
    type: RollingUpdate
```

**Comment by NickLarsenNZ** [comment]
IMO, there are too many additional steps to make this work.
I think this should be reopened, and have the image-updater support ECR (and other major registries).

**Comment by mrobinson-wavehq** [comment]
> IMO, there are too many additional steps to make this work. @jannfis I think this should be reopened, and have the image-updater support ECR (and other major registries).

I'd like to +1 this. My org is building out a container platform in AWS involving EKS with heavy use of Argo projects. As mentioned in other comments, supporting [IRSA](https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts.html) would be ideal to allow least privilege access; it's our de-facto method for pods to interact with the AWS API and has worked well for us.
We're far enough out from going to production that we could contribute testing as needed.

**Comment by jannfis** [comment]
The aws CLI is part of the image updater's container image since a while, so setting up authentication against ECR should be quite simple. As I'm not an aws user myself, there's not really much I can (or want) to do here. I'll happily accept contributions from the community to make this easier, but I'll not be able to support those (for the previously mentioned reasons).

**Comment by mubarak-j** [comment]
I wanted to point out that most of these steps are now simplified by the [helm chart](https://github.com/argoproj/argo-helm/blob/master/charts/argocd-image-updater/values.yaml). e.g:
[registries.com](https://github.com/argoproj/argo-helm/blob/aa418962922d5967aaebe7a7fb362fa21d1104e2/charts/argocd-image-updater/values.yaml#L89-L95) can be configured as follow:
```yaml
config:
  registries:
  - name: Docker Hub
    api_url: https://registry-1.docker.io
    prefix: docker.io
  - name: ECR
    api_url: https://xxx.dkr.ecr.us-west-1.amazonaws.com
    prefix: ""
    ping: yes
    insecure: no
    credentials: ext:/scripts/ecr-login.sh.  #script name should match here and in authScripts 
    credsexpire: 11h
```

and configure [ECR authentication script](https://github.com/argoproj/argo-helm/blob/aa418962922d5967aaebe7a7fb362fa21d1104e2/charts/argocd-image-updater/values.yaml#L99-L109) as follow:
```yaml
authScripts:
  enabled: true
  scripts: 
    ecr-login.sh: |   # notice script name matches above    
      #!/bin/sh
      aws ecr --region $AWS_REGION get-authorization-token --output text --query 'authorizationData[].authorizationToken' | base64 -d
```
In my case, `$AWS_REGION` is set as [environment variable](https://github.com/argoproj/argo-helm/blob/d30d85060e31b2bacc26afa01a00787950ed25b0/charts/argocd-image-updater/values.yaml#L37-L40).
```yaml
extraEnv:
  - name: AWS_REGION
    value: "us-west-1"
```
This setup will mount the login script as `/scripts/ecr-login.sh` and will run every 11 hours.

I'm able to run image-updater on EKS with IRSA configured. Again using the current helm version, you will need to enable [service account for image-updater](https://github.com/argoproj/argo-helm/blob/d30d85060e31b2bacc26afa01a00787950ed25b0/charts/argocd-image-updater/values.yaml#L113-L120) and assign a role with appropriate IAM permissions, e.g: 

```yaml
serviceAccount:
  create: true
  annotations:
    eks.amazonaws.com/role-arn: arn:aws:iam::<Account-ID>:role/<role-name>
  name: "argocd-image-updater"  # I think this is the default
```

For IAM permissions you will at least need [AmazonEC2ContainerRegistryReadOnly](https://docs.aws.amazon.com/AmazonECR/latest/userguide/security-iam-awsmanpol.html#security-iam-awsmanpol-AmazonEC2ContainerRegistryReadOnly) and if you're new to IRSA [this docs](https://shipit.dev/posts/setting-up-eks-with-irsa-using-terraform.html) might help you get started. 

I hope this helps others get started with image-updater on AWS. 
@jannfis if you think this is worth adding to the docs, im happy to create a PR.

**Comment by paul-benetis** [comment]
@mubarak-j While this setup will work with image-updater to update images for already existing applications, will this work for authenticating argocd itself for the initial deploy of the application?

**Comment by rishi-cloudhero** [comment]
ConfigMap for ECR authentication
```
apiVersion: v1
kind: ConfigMap
metadata:
  labels:
    app.kubernetes.io/name: argocd-image-updater-config
    app.kubernetes.io/part-of: argocd-image-updater
  name: argocd-image-updater-config
  namespace: argocd
data:
  registries.conf: |
    registries:
    - name: AWS ECR
      prefix: <aws_account_id>.dkr.ecr.eu-west-2.amazonaws.com
      api_url: https://<aws_account_id>.dkr.ecr.eu-west-2.amazonaws.com
      credentials: secret:argocd/ecr-secrets#creds
      default: true
      insecure: yes
     credsexpire: 12h
      
```
for authenticating to ecr, use the cronjob to generate credentials on every schedule.

**Comment by rraj-gautam** [comment]
ConfigMap for ECR authentication
```
apiVersion: v1
kind: ConfigMap
metadata:
  labels:
    app.kubernetes.io/name: argocd-image-updater-config
    app.kubernetes.io/part-of: argocd-image-updater
  name: argocd-image-updater-config
  namespace: argocd
data:
  registries.conf: |
    registries:
    - name: AWS ECR
      prefix: <aws_account_id>.dkr.ecr.eu-west-2.amazonaws.com
      api_url: https://<aws_account_id>.dkr.ecr.eu-west-2.amazonaws.com
      credentials: secret:argocd/ecr-secrets#creds
      default: true
      insecure: yes
     credsexpire: 12h
```      
for authenticating to ecr, use the cronjob to generate credentials on every schedule.

**Comment by joebowbeer** [comment]
I've been using [keel.sh](https://keel.sh) partly because of its excellent ECR integration via IRSA and an AWS-aware credentials helper:

https://github.com/keel-hq/keel/tree/master/extension/credentialshelper/aws

But it would me nice to have another alternative.

**Comment by vikas027** [comment]
> I've been using [keel.sh](https://keel.sh) partly because of its excellent ECR integration via IRSA and an AWS-aware credentials helper:
> 

Keel hasn't updated for a while, looks like the project has been abandoned :(

**Comment by joebowbeer** [comment]
> Keel hasn't updated for a while, looks like the project has been abandoned :(

I can't disagree but it still functions. keel is my only option at the moment, and if it breaks then someone may need to fork it and fix it. It would be nice to have another option..

**Comment by vikas027** [comment]
> keel is my only option at the moment

I have not tried Image Updater yet but there are few people who have successfully got the updater working with ECR. Did you try that?

**Comment by joebowbeer** [comment]
@vikas027 wrote:
> there are few people who have successfully got the updater working with ECR

Please interpret my comments as an RFE for ECR integration via IRSA and an AWS-aware credentials helper.

**Comment by fabioaraujopt** [comment]
> I wanted to point out that most of these steps are now simplified by the [helm chart](https://github.com/argoproj/argo-helm/blob/master/charts/argocd-image-updater/values.yaml). e.g: [registries config](https://github.com/argoproj/argo-helm/blob/aa418962922d5967aaebe7a7fb362fa21d1104e2/charts/argocd-image-updater/values.yaml#L89-L95) can be configured as follow:
> 
> ```yaml
> config:
>   registries:
>   - name: Docker Hub
>     api_url: https://registry-1.docker.io
>     prefix: docker.io
>   - name: ECR
>     api_url: https://xxx.dkr.ecr.us-west-1.amazonaws.com
>     prefix: xxx.dkr.ecr.us-west-1.amazonaws.com   # before v0.12 this needed to be set empty, prefix: ""
>     default: true  # not supported before v0.12
>     ping: yes
>     insecure: no
>     credentials: ext:/scripts/ecr-login.sh  #script name should match here and in authScripts 
>     credsexpire: 11h
> ```
> 
> and configure [ECR authentication script](https://github.com/argoproj/argo-helm/blob/aa418962922d5967aaebe7a7fb362fa21d1104e2/charts/argocd-image-updater/values.yaml#L99-L109) as follow:
> 
> ```yaml
> authScripts:
>   enabled: true
>   scripts: 
>     ecr-login.sh: |   # notice script name matches above    
>       #!/bin/sh
>       aws ecr --region $AWS_REGION get-authorization-token --output text --query 'authorizationData[].authorizationToken' | base64 -d
> ```
> 
> In my case, `$AWS_REGION` is set as [environment variable](https://github.com/argoproj/argo-helm/blob/d30d85060e31b2bacc26afa01a00787950ed25b0/charts/argocd-image-updater/values.yaml#L37-L40).
> 
> ```yaml
> extraEnv:
>   - name: AWS_REGION
>     value: "us-west-1"
> ```
> 
> This setup will mount the login script as `/scripts/ecr-login.sh` and will run every 11 hours.
> 
> I'm able to run image-updater on EKS with IRSA configured. Again using the current helm version, you will need to enable [service account for image-updater](https://github.com/argoproj/argo-helm/blob/d30d85060e31b2bacc26afa01a00787950ed25b0/charts/argocd-image-updater/values.yaml#L113-L120) and assign a role with appropriate IAM permissions, e.g:
> 
> ```yaml
> serviceAccount:
>   create: true
>   annotations:
>     eks.amazonaws.com/role-arn: arn:aws:iam::<Account-ID>:role/<role-name>
>   name: "argocd-image-updater"  # I think this is the default
> ```
> 
> For IAM permissions you will at least need [AmazonEC2ContainerRegistryReadOnly](https://docs.aws.amazon.com/AmazonECR/latest/userguide/security-iam-awsmanpol.html#security-iam-awsmanpol-AmazonEC2ContainerRegistryReadOnly).
> 
> I hope this helps others get started with image-updater on AWS. @jannfis if you think this is worth adding to the docs, im happy to create a PR.

I'm having this error:

Could not set registry endpoint credentials: error executing /scripts/ecr-login.sh: fork/exec /scripts/ecr-login.sh: no such file or directory" 

I'm using the following config: 

`
config:
  registries:
    - name: ECR
      api_url: https://XXX.dkr.ecr.eu-west-1.amazonaws.com
      prefix: XXX.dkr.ecr.eu-west-1.amazonaws.com
      ping: yes
      insecure: no
      credentials: ext:/scripts/ecr-login.sh
      credsexpire: 6h
authScripts:
  enabled: true
  scripts:
    ecr-login.sh: |
      #!/bin/sh
      aws ecr get-authorization-token --region eu-west-1 --registry-ids XXXX --output text --query 'authorizationData[].authorizationToken'
`

**Comment by thuandt** [comment]
@fabioaraujopt Please ensure `ecr-login.sh` script exist in argocd-image-updater container (mount as ConfigMap in my case)

**Comment by fmubaidien** [comment]
currently facing this issue:

**Comment by fmubaidien** [comment]
currently facing this issue when I run the test:

**Comment by fmubaidien** [comment]
Currently facing this issue:
`FATA[0000] could not set registry credentials: error executing /scripts/ecr-login.sh: `/scripts/ecr-login.sh` failed exit status 255: An error occurred (AccessDenied) when calling the AssumeRoleWithWebIdentity operation: Not authorized to perform sts:AssumeRoleWithWebIdentity  application=test image_alias= image_name=xxx.dkr.ecr.eu-west-2.amazonaws.com/catches-api registry_url=xxx.dkr.ecr.eu-west-2.amazonaws.com
~/config $`
Any ideas what could be the reason? service account and role have been created and connected

**Comment by hoangduy1912** [comment]
> I wanted to point out that most of these steps are now simplified by the [helm chart](https://github.com/argoproj/argo-helm/blob/master/charts/argocd-image-updater/values.yaml). e.g: [registries config](https://github.com/argoproj/argo-helm/blob/aa418962922d5967aaebe7a7fb362fa21d1104e2/charts/argocd-image-updater/values.yaml#L89-L95) can be configured as follow:
> 
> ```yaml
> config:
>   registries:
>   - name: Docker Hub
>     api_url: https://registry-1.docker.io
>     prefix: docker.io
>   - name: ECR
>     api_url: https://xxx.dkr.ecr.us-west-1.amazonaws.com
>     prefix: xxx.dkr.ecr.us-west-1.amazonaws.com   # before v0.12 this needed to be set empty, prefix: ""
>     default: true  # not supported before v0.12
>     ping: yes
>     insecure: no
>     credentials: ext:/scripts/ecr-login.sh  #script name should match here and in authScripts 
>     credsexpire: 11h
> ```
> 
> and configure [ECR authentication script](https://github.com/argoproj/argo-helm/blob/aa418962922d5967aaebe7a7fb362fa21d1104e2/charts/argocd-image-updater/values.yaml#L99-L109) as follow:
> 
> ```yaml
> authScripts:
>   enabled: true
>   scripts: 
>     ecr-login.sh: |   # notice script name matches above    
>       #!/bin/sh
>       aws ecr --region $AWS_REGION get-authorization-token --output text --query 'authorizationData[].authorizationToken' | base64 -d
> ```
> 
> In my case, `$AWS_REGION` is set as [environment variable](https://github.com/argoproj/argo-helm/blob/d30d85060e31b2bacc26afa01a00787950ed25b0/charts/argocd-image-updater/values.yaml#L37-L40).
> 
> ```yaml
> extraEnv:
>   - name: AWS_REGION
>     value: "us-west-1"
> ```
> 
> This setup will mount the login script as `/scripts/ecr-login.sh` and will run every 11 hours.
> 
> I'm able to run image-updater on EKS with IRSA configured. Again using the current helm version, you will need to enable [service account for image-updater](https://github.com/argoproj/argo-helm/blob/d30d85060e31b2bacc26afa01a00787950ed25b0/charts/argocd-image-updater/values.yaml#L113-L120) and assign a role with appropriate IAM permissions, e.g:
> 
> ```yaml
> serviceAccount:
>   create: true
>   annotations:
>     eks.amazonaws.com/role-arn: arn:aws:iam::<Account-ID>:role/<role-name>
>   name: "argocd-image-updater"  # I think this is the default
> ```
> 
> For IAM permissions you will at least need [AmazonEC2ContainerRegistryReadOnly](https://docs.aws.amazon.com/AmazonECR/latest/userguide/security-iam-awsmanpol.html#security-iam-awsmanpol-AmazonEC2ContainerRegistryReadOnly).
> 
> I hope this helps others get started with image-updater on AWS. @jannfis if you think this is worth adding to the docs, im happy to create a PR.

Hi, I'm trying to configure as mentioned here but still encountered this error

ERRO[0000] Could not get tags from registry: Get "https://awsNumber.dkr.ecr.eu-central-1.amazonaws.com/v2/projectName/tags/list": no basic auth credentials  alias= application=projectName image_name=ProjectName image_tag=ui-latest registry=awsNumber.dkr.ecr.eu-central-1.amazonaws.com
INFO[0000] Finished cache warm-up, pre-loaded 0 meta data entries from 2 registries
ERRO[0000] Health probe server exited with error: listen tcp :8080: bind: address already in use


Here's what I'm using
`
config:
  registries:
  - name: ECR
    api_url: https://awsNumber.dkr.ecr.eu-central-1.amazonaws.com
    prefix: awsNumber.dkr.ecr.eu-central-1.amazonaws.com   # before v0.12 this needed to be set empty, prefix: ""
    default: true  # not supported before v0.12
    ping: yes
    insecure: no
    credentials: ext:/scripts/ecr-login.sh  #script name should match here and in authScripts 
    credsexpire: 11h
authScripts:
  enabled: true
  scripts: 
    ecr-login.sh: |   # notice script name matches above    
      #!/bin/sh
      aws ecr --region $AWS_REGION get-authorization-token --output text --query 'authorizationData[].authorizationToken' | base64 -d
extraEnv:
  - name: AWS_REGION
    value: "eu-central-1"
serviceAccount:
  create: true
  annotations:
    eks.amazonaws.com/role-arn: arn:aws:iam::awsNumber:role/eksWorker
  name: "argocd-image-updater"
# config:
#   logLevel: debug
metrics:
  serviceMonitor:
    namespace: argocd
`

**Comment by hoangduy1912** [comment]
Hi everyone, I'm encountering this error

Below is argocd-image-updater pod log
```

time="2022-12-05T09:33:14Z" level=info msg="Starting image update cycle, considering 1 annotated application(s) for update"
time="2022-12-05T09:33:14Z" level=debug msg="Processing application ui-connectivity"
time="2022-12-05T09:33:14Z" level=debug msg="Considering this image for update" alias= application=ui-connectivity image_name=qa image_tag=latest registry=<aws-number>.dkr.ecr.eu-central-1.amazonaws.com
time="2022-12-05T09:33:14Z" level=debug msg="Using version constraint 'latest' when looking for a new tag" alias= application=ui-connectivity image_name=qa image_tag=latest registry=<aws-number>.dkr.ecr.eu-central-1.amazonaws.com
time="2022-12-05T09:33:14Z" level=info msg=/scripts/ecr-login.sh dir= execID=657b6
time="2022-12-05T09:33:15Z" level=error msg="Could not set registry endpoint credentials: invalid script output, must be single line with syntax <username>:<password>" alias= application=ui-connectivity image_name=qa image_tag=latest registry=<aws-number>.dkr.ecr.eu-central-1.amazonaws.com
time="2022-12-05T09:33:15Z" level=info msg="Processing results: applications=1 images_considered=1 images_skipped=0 images_updated=0 errors=1"

```

When I execute the script within the pod, the following error was returned
```
/ $ ./scripts/ecr-login.sh
An error occurred (AccessDenied) when calling the AssumeRoleWithWebIdentity operation: Not authorized to perform sts:AssumeRoleWithWebIdentity
```

I have checked that the pod is using correct service account. Policy AmazonEC2ContainerRegistryReadOnly was added to that AWS role

Trust relationship added

```
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Sid": "",
            "Effect": "Allow",
            "Principal": {
                "Service": "ec2.amazonaws.com"
            },
            "Action": [
                "sts:AssumeRole",
                "sts:AssumeRoleWithWebIdentity"
            ]
        }
    ]
}
```

**Comment by zrice57** [comment]
@fabioaraujopt I was having the same issue. I believe the version of the chart which I was installing, `0.8.1`, by default uses the same image version and that image version does not contain any `sh` or `bash` executables. It is probably the lack of those files resulting in `no such file or directory`.

I was able to get it working by specifying the latest image version in my `values` [file](https://github.com/argoproj/argo-helm/blob/main/charts/argocd-image-updater/values.yaml#L3): 

```
image:
  tag: "v0.12.1"
```

I'm not sure how other people were getting this to work, perhaps we are using different versions of the helm chart? `0.8.1` appears to be the latest version.

**Comment by kyleli666** [comment]
Hi @mubarak-j, I followed your flow with the latest helm chart, but I cannot see any logs mentioning the auth1.sh script is called.
However, I can manually run the script correctly from inside the container, and get the credential.
What could be the problem?
Here is the log: 
```
time="2023-07-21T03:30:12Z" level=info msg="argocd-image-updater v0.12.2+1aa317c starting [loglevel:TRACE, interval:2m0s, healthport:8080]"
time="2023-07-21T03:30:12Z" level=warning msg="commit message template at /app/config/commit.template does not exist, using default"
time="2023-07-21T03:30:12Z" level=debug msg="Successfully parsed commit message template"
time="2023-07-21T03:30:12Z" level=debug msg="rate limiting is disabled" prefix=14351427915.dkr.ecr.ap-southeast-1.amazonaws.com registry="https://14351427915.dkr.ecr.ap-southeast-1.amazonaws.com"
time="2023-07-21T03:30:12Z" level=debug msg="Setting default registry endpoint to 14351427915.dkr.ecr.ap-southeast-1.amazonaws.com"
time="2023-07-21T03:30:12Z" level=debug msg="Previous default registry was docker.io"
time="2023-07-21T03:30:12Z" level=info msg="Loaded 1 registry configurations from /app/config/registries.conf"
time="2023-07-21T03:30:12Z" level=debug msg="Creating in-cluster Kubernetes client"
time="2023-07-21T03:30:12Z" level=info msg="ArgoCD configuration: [apiKind=kubernetes, server=argocd-server.argocd, auth_token=false, insecure=false, grpc_web=true, plaintext=false]"
time="2023-07-21T03:30:12Z" level=info msg="Starting health probe server TCP port=8080"
time="2023-07-21T03:30:12Z" level=info msg="Starting metrics server on TCP port=8081"
time="2023-07-21T03:30:12Z" level=info msg="Warming up image cache"
time="2023-07-21T03:30:12Z" level=trace msg="skipping app 'argocd-image-updater' of type '' because required annotation is missing" application=argocd-image-updater
time="2023-07-21T03:30:12Z" level=trace msg="processing app 'ba-deployment-api' of type 'Helm'" application=ba-deployment-api
time="2023-07-21T03:30:12Z" level=debug msg="Processing application ba-deployment-api"
time="2023-07-21T03:30:12Z" level=debug msg="Considering this image for update" alias=app application=ba-deployment-api image_name=ba.deployment.api image_tag=30 registry=1234567890.dkr.ecr.ap-southeast-1.amazonaws.com
time="2023-07-21T03:30:12Z" level=debug msg="setting rate limit to 20 requests per second" prefix=1234567890.dkr.ecr.ap-southeast-1.amazonaws.com registry="https://1234567890.dkr.ecr.ap-southeast-1.amazonaws.com"
time="2023-07-21T03:30:12Z" level=debug msg="Inferred registry from prefix 1234567890.dkr.ecr.ap-southeast-1.amazonaws.com to use API https://1234567890.dkr.ecr.ap-southeast-1.amazonaws.com"
time="2023-07-21T03:30:12Z" level=debug msg="Using no version constraint when looking for a new tag" alias=app application=ba-deployment-api image_name=ba.deployment.api image_tag=30 registry=1234567890.dkr.ecr.ap-southeast-1.amazonaws.com
time="2023-07-21T03:30:12Z" level=trace msg="Found update strategy latest" image_alias=app image_name=1234567890.dkr.ecr.ap-southeast-1.amazonaws.com/ba.deployment.api registry_url=1234567890.dkr.ecr.ap-southeast-1.amazonaws.com
time="2023-07-21T03:30:12Z" level=trace msg="No match annotation found" image_alias=app image_name=1234567890.dkr.ecr.ap-southeast-1.amazonaws.com/ba.deployment.api registry_url=1234567890.dkr.ecr.ap-southeast-1.amazonaws.com
time="2023-07-21T03:30:12Z" level=trace msg="No ignore-tags annotation found" image_alias=app image_name=1234567890.dkr.ecr.ap-southeast-1.amazonaws.com/ba.deployment.api registry_url=1234567890.dkr.ecr.ap-southeast-1.amazonaws.com
time="2023-07-21T03:30:12Z" level=trace msg="Using runtime platform constraint linux/amd64" image_alias=app image_name=1234567890.dkr.ecr.ap-southeast-1.amazonaws.com/ba.deployment.api registry_url=1234567890.dkr.ecr.ap-southeast-1.amazonaws.com
time="2023-07-21T03:30:12Z" level=trace msg="No pull-secret annotation found" image_alias=app image_name=1234567890.dkr.ecr.ap-southeast-1.amazonaws.com/ba.deployment.api registry_url=1234567890.dkr.ecr.ap-southeast-1.amazonaws.com
time="2023-07-21T03:30:12Z" level=trace msg="Performing HTTP GET https://1234567890.dkr.ecr.ap-southeast-1.amazonaws.com/v2/ba.deployment.api/tags/list"
time="2023-07-21T03:30:12Z" level=error msg="Could not get tags from registry: Get \"https://1234567890.dkr.ecr.ap-southeast-1.amazonaws.com/v2/ba.deployment.api/tags/list\": no basic auth credentials" alias=app application=ba-deployment-api image_name=ba.deployment.api image_tag=30 registry=1234567890.dkr.ecr.ap-southeast-1.amazonaws.com
time="2023-07-21T03:30:12Z" level=info msg="Finished cache warm-up, pre-loaded 0 meta data entries from 3 registries"
```

**Comment by zhaque44** [comment]
> ConfigMap for ECR authentication
> 
> ```
> apiVersion: v1
> kind: ConfigMap
> metadata:
>   labels:
>     app.kubernetes.io/name: argocd-image-updater-config
>     app.kubernetes.io/part-of: argocd-image-updater
>   name: argocd-image-updater-config
>   namespace: argocd
> data:
>   registries.conf: |
>     registries:
>     - name: AWS ECR
>       prefix: <aws_account_id>.dkr.ecr.eu-west-2.amazonaws.com
>       api_url: https://<aws_account_id>.dkr.ecr.eu-west-2.amazonaws.com
>       credentials: secret:argocd/ecr-secrets#creds
>       default: true
>       insecure: yes
>      credsexpire: 12h
> ```
> 
> for authenticating to ecr, use the cronjob to generate credentials on every schedule.

do you have an example repo for this?

**Comment by mccullya** [comment]
The scripts above i.e. 
`aws ecr get-authorization-token --region eu-west-1 --registry-ids XXXX --output text --query 'authorizationData[].authorizationToken'`

were not working for me.  It was erroring saying that it wants a username:password.  I used this instead, and it works :)

```
authScripts:
  enabled: true
  scripts: 
    ecr-login.sh: |
      #!/bin/sh
      # Retrieve the authorization token from AWS ECR
      auth_token=$(aws ecr get-authorization-token --region eu-west-1 --output text --query 'authorizationData[].authorizationToken')
      
      # Decode the authorization token
      decoded_token=$(echo $auth_token | base64 -d)
      
      # Extract username and password
      username=$(echo $decoded_token | cut -d: -f1)
      password=$(echo $decoded_token | cut -d: -f2)
      
      # Output username and password
      echo "$username:$password"
```

**Comment by bradenwright** [comment]
Fwiw I followed this article and it worked for me, https://medium.com/@tomas94depi/argo-image-updater-with-aws-ecr-ddb661abb332

**Comment by giogujabidze** [comment]
Hi guys, I'm afraid this workaround/solution stoped working in a recent versions. As currently I'm using the helm chart version **`0.11.0`** which uses the application **`v0.14.0`** and the script does not work, I guess because of the fact, that the /app directory is mounted as a readonly, to whenever I try to run aws commands from within the pod, it throws the following error message: `[Errno 30] Read-only file system: '/app/.aws'`

I tried to solve that by using the custom credentials files using aws reserved environment variables:
```yaml
  extraEnv:
    - name: AWS_REGION
      value: "us-east-1"
    - name: AWS_CONFIG_FILE
      value: "/tmp/.aws/config"
    - name: AWS_SHARED_CREDENTIALS_FILE
      value: "/tmp/.aws/credentials"
```
After this, if I run 'aws configure' it creates _config_ file under `/tmp/.aws` directory, but if I run for example 'aws ecr get-authorization-token --region eu-west-1' commnad, it still tries to use the `$HOME/.aws` directory (which refers to /apps/.aws) and still fails. 

Maybe your container image has somehow hardcoded aws configuration file location?

This is my entire configuration (helm value file):

```yaml
  config:
    applicationsAPIKind: "kubernetes"

    logLevel: "debug"

    registries:
    - name: ECR
      api_url: https://<ACCOUNT_ID>.dkr.ecr.us-east-1.amazonaws.com
      prefix: <ACCOUNT_ID>.dkr.ecr.us-east-1.amazonaws.com
      default: true
      ping: yes
      insecure: no
      credentials: ext:/scripts/ecr-login.sh
      credsexpire: 10h

  extraEnv:
    - name: AWS_REGION
      value: "us-east-1"
    - name: AWS_CONFIG_FILE
      value: "/tmp/.aws/config"
    - name: AWS_SHARED_CREDENTIALS_FILE
      value: "/tmp/.aws/credentials"

  serviceAccount:
    create: true
    annotations: 
      eks.amazonaws.com/role-arn: arn:aws:iam::464682684796:role/argocd-image-updater-role
    name: "argocd-image-updater"

  # whether to mount authentication scripts, if enabled, the authentication scripts will be mounted on /scripts that can be used to authenticate with registries (ECR)
  # refer to https://argocd-image-updater.readthedocs.io/en/stable/configuration/registries/#specifying-credentials-for-accessing-container-registries for more info
  authScripts:
    # -- Whether to mount the defined scripts that can be used to authenticate with a registry, the scripts will be mounted at `/scripts`
    enabled: true
    scripts:
      ecr-login.sh: |
        #!/bin/sh
        echo "AWS:$(aws ecr get-login-password --region us-east-1)"
```

P.S. I tried to use the script described [above](https://github.com/argoproj-labs/argocd-image-updater/issues/112#issuecomment-1058465942):
`aws ecr --region $AWS_REGION get-authorization-token --output text --query 'authorizationData[].authorizationToken' | base64 -d`
but faced the issue which already is described [here](https://github.com/argoproj-labs/argocd-image-updater/issues/112#issuecomment-1886870990), so found the simpler command (since the username is always the same):
`echo "AWS:$(aws ecr get-login-password --region us-east-1)"`
But like I said, i cannot test/run any awscli comands from within the argocd-image-updater pod. (((

**Comment by mubarak-j** [comment]
@giogujabidze it seems the breaking change was introduced in [helm chart v0.10.0](https://github.com/argoproj/argo-helm/pull/2692/files) which added `securityContext` with default value `readOnlyRootFilesystem: true`, you may want to override that in your helm values.

**Comment by giogujabidze** [comment]
> @giogujabidze it seems the breaking change was introduced in [helm chart v0.10.0](https://github.com/argoproj/argo-helm/pull/2692/files) which added `securityContext` with default value `readOnlyRootFilesystem: true`, you may want to override that in your helm values.

Thanks @mubarak-j for confirming my doubts. 
Adding the following section to my helm values file fix everything:
```yaml
securityContext:
  readOnlyRootFilesystem: false
  ```

**Comment by giogujabidze** [comment]
I'm putting the whole helm values file content for others, who will try to use ArgoCD Image Updater against AWS ECR, to not lose as much time as I did to make it work. Here is the configuration that worked for me (helm chart version - `0.11.0`):

```yaml
extraEnv:
  - name: AWS_REGION
    value: "us-east-1"

config:
  applicationsAPIKind: "kubernetes"
  logLevel: "debug"

  registries:
  - name: ECR
    api_url: https://<AWS_ACCOUNT>.dkr.ecr.us-east-1.amazonaws.com
    prefix: <AWS_ACCOUNT>.dkr.ecr.us-east-1.amazonaws.com
    default: true
    ping: yes
    insecure: no
    credentials: ext:/scripts/ecr-login.sh
    credsexpire: 10h
 
authScripts:
  enabled: true
  scripts:
    ecr-login.sh: |
      #!/bin/sh
      echo "AWS:$(aws ecr get-login-password --region $AWS_REGION)"

serviceAccount:
  create: true
  annotations: 
    eks.amazonaws.com/role-arn: arn:aws:iam::<AWS_ACCOUNT>:role/argocd-image-updater-role
  name: "argocd-image-updater"

securityContext:
  readOnlyRootFilesystem: false
```

**Comment by peter-nguyen-rw** [comment]
Hello. I'm still running into this issue regarding : no basic auth credentials

@giogujabidze I have the same exact Helm fields as you posted above. Any ideas?

**Comment by giogujabidze** [comment]
> Hello. I'm still running into this issue regarding : no basic auth credentials
> 
> @giogujabidze I have the same exact Helm fields as you posted above. Any ideas?

Hi. Well it's difficult to say exact reason, as I also spent a lot of time to make it work, but in your case I would check the IRSA policy, maybe you missed some essential permissions and because of that, you cannot actually authenticate against ECR. [Here](https://docs.aws.amazon.com/AmazonECR/latest/userguide/common-errors-docker.html#error-403) is the list of some possible reasons for this error message.

**Comment by peter-nguyen-rw** [comment]
@giogujabidze 
Do you mind if I see your IRSA policy and the the policy for the role associated with it? Also can you tell me what Helm targetRevision and image version you're on for the Image Updater please?

**Comment by giogujabidze** [comment]
> @giogujabidze Do you mind if I see your IRSA policy and the the policy for the role associated with it? Also can you tell me what Helm targetRevision and image version you're on for the Image Updater please?

This is the policy attached to my IRSA role for image updater:
```json
{
    "Statement": [
        {
            "Action": [
                "ecr:ValidatePullThroughCacheRule",
                "ecr:ListTagsForResource",
                "ecr:ListImages",
                "ecr:GetRepositoryPolicy",
                "ecr:GetRegistryScanningConfiguration",
                "ecr:GetLifecyclePolicy",
                "ecr:GetDownloadUrlForLayer",
                "ecr:GetAuthorizationToken",
                "ecr:DescribeRepositories",
                "ecr:DescribeRegistry",
                "ecr:DescribePullThroughCacheRules",
                "ecr:DescribeImages",
                "ecr:DescribeImageScanFindings",
                "ecr:BatchGetRepositoryScanningConfiguration",
                "ecr:BatchGetImage",
                "ecr:BatchCheckLayerAvailability"
            ],
            "Effect": "Allow",
            "Resource": "*",
            "Sid": ""
        }
    ],
    "Version": "2012-10-17"
}
```
As for the helm targetRevision, like I already mentioned above, I am using the latest (at this moment) version of helm chart - `0.11.0`

**Comment by peter-nguyen-rw** [comment]
@giogujabidze 
Okay so this is odd. Because the 0.11.0 version of the Helm chart apparently does not support the field `default: true` within your registries.conf. 

See this comment here about that field not being supported for versions before 0.12.0
https://github.com/argoproj-labs/argocd-image-updater/issues/112#issuecomment-1058465942

**Comment by giogujabidze** [comment]
> @giogujabidze Okay so this is odd. Because the 0.11.0 version of the Helm chart apparently does not support the field `default: true` within your registries.conf.
> 
> See this comment here about that field not being supported for versions before 0.12.0 [#112 (comment)](https://github.com/argoproj-labs/argocd-image-updater/issues/112#issuecomment-1058465942)

I believe that comment was referred to the application version itself, as the helm chart `0.11.0` includes the application version `v0.14.0`- https://github.com/argoproj/argo-helm/blob/main/charts/argocd-image-updater/Chart.yaml#L6

That means that if you'll deploy this helm chart, it will deploy the argocd image updater docker image version v0.14.0 by default - https://github.com/argoproj/argo-helm/blob/main/charts/argocd-image-updater/templates/deployment.yaml#L130

Anyways, like I said, it works in my case and it's difficult to say, why it's not working for you.

**Comment by peter-nguyen-rw** [comment]
@giogujabidze 
Okay thanks for that clarification.

Sorry, back to your IAM setup. The policy you pasted to me is directly attached to your IRSA role? You're not using the IRSA role to assume a role that has those policies attached to it?

**Comment by giogujabidze** [comment]
Hey, sorry for late reply

> The policy you pasted to me is directly attached to your IRSA role?

Yes, the IAM policy is attached to the role, arn of which is set in annotations of service account for image-updater

**Comment by chuong-dao** [comment]
Looks like my setup can see my registry and tags but when I update a new docker image, it's still showing the same message "no newer version of image found "
Not sure what to do at this point.

argocd-image-updater test XXXXXXXXXX.dkr.ecr.us-east-2.amazonaws.com/argocd-updater-test --registries-conf-path=/app/config/registries.conf
```
INFO[0000] Found 2 tags in registry                      application=test image_alias= image_name=XXXXXXXXXX.dkr.ecr.us-east-2.amazonaws.com/argocd-updater-test registry_url=XXXXXXXXXX.dkr.ecr.us-east-2.amazonaws.com
DEBU[0000] could not parse input tag latest as semver: Invalid Semantic Version
DEBU[0000] could not parse input tag latest-v1 as semver: Invalid Semantic Version
INFO[0000] no newer version of image found               application=test image_alias= image_name=XXXXXXXXXX.dkr.ecr.us-east-2.amazonaws.com/argocd-updater-test registry_url=XXXXXXXXXX.dkr.ecr.us-east-2.amazonaws.com
```

**Comment by Slhttnunlu** [comment]
Greetings, 

I am trying to auth to ECR repo with IRSA, I created a role using this policy [https://docs.aws.amazon.com/AmazonECR/latest/userguide/security-iam-awsmanpol.html#security-iam-awsmanpol-AmazonEC2ContainerRegistryReadOnly] and added trusted policy, and following the examples in the comments I added the role I created to argocd-image-updater serviceaccount as arn, but when I restart my pod I still get auth error, 

When I use the auth method with pullsecret, everything goes smoothly, but for 12 hours, can anyone successfully use ECR auth with IAM role?

**Comment by washanhanzi** [comment]
For me, the annotation on the service account won't work. I need to manually create the service account use `eksctl` cli, [here](https://docs.aws.amazon.com/eks/latest/userguide/associate-service-account-role.html#_step_2_create_and_associate_iam_role). 

The policy is same as previous comment.

**Comment by ritik872000** [comment]
Hi Folks, I have written this article on how to make it all work , the article consists the steps which I have taken to fix the issue.
https://medium.com/@ritikkesharwani3/aws-ecr-argocd-automate-helm-chart-access-with-cross-account-token-rotation-582efe20a748
